wikimedia/wikipedia-ios
# Add Featured article widget

**Phabricator:** https://phabricator.wikimedia.org/T259842

### Notes

This PR adds the Featured article widget and an alternative data retrieval strategy for that widget. Instead of directly accessing our shared Core Data store in the widget extension, we're directly accessing Wikifeeds in the widget extension to pull featured content and caching that in a small `WidgetCache` on the file system. In the primary app, when the user changes their preferred language, we're persisting that language change into the this same `WidgetCache` for the widget extension to consume.  

This retrieval strategy is currently only implemented for the Featured article widget and for this reason was kept somewhat isolated in `WidgetController`. If we apply it to the other widgets, we should refactor `WidgetController` as mentioned in the comments. The rationale behind this retrieval strategy is available in more detail here: https://phabricator.wikimedia.org/T268677#7168322

There's a bit more polish work to do [pending responses](https://phabricator.wikimedia.org/T259842#7133115) – I'll include them as a follow up PR after those questions are answered.

### Test Steps

1. Add a Featured article widget of each size class
2. Set your primary Wikipedia language to English, observe that the widget loads the featured article of the day
3. After midnight every day, the widget should perform a Wikifeeds request and update with a new article if available 
4. Set your primary Wikipedia language to an unsupported language (like Spanish), and confirm an error message is displayed on the widget
5. To observe the alternate design of the widget (where a lead image is unavailable for an article), change `imageData` [here](https://github.com/wikimedia/wikipedia-ios/pull/3988/files#diff-f22d1151671dec30edd37cc62d1f6acb7b81b149df35b8e63ccfc33e591cff95R69) to return `nil`.

There still may be some cracks in the system, so if possible I'd recommend adding the widget and casually observing it over the course of a few days. One particular language I noticed issues with during development was German having problems loading content, but haven't yet been able to replicate this as the widget was completed, so I'm unclear if it's a widget issue or feed issue - so this may be one language worth observing as well.
